nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -312,100 +342,236 @@ func (w *deploymentWatcher) StopWatch() {
 	w.exitFn()
 }
 
-// watch is the long running watcher that takes actions upon allocation changes
+// watch is the long running watcher that watches for both allocation and
+// deployment changes. Its function is to create evaluations to trigger the
+// scheduler when more progress can be made, to fail the deployment if it has
+// failed and potentially rolling back the job.
 func (w *deploymentWatcher) watch() {
+	// Get the deadline. This is likely a zero time to begin with but we need to
+	// handle the case that the deployment has already progressed and we are now
+	// just starting to watch it.
+	currentDeadline := getDeploymentProgressCutoff(w.getDeployment())
+	var deadlineTimer *time.Timer
+	if currentDeadline.IsZero() {
+		deadlineTimer = time.NewTimer(0)
+		if !deadlineTimer.Stop() {
+			<-deadlineTimer.C
+		}
+	} else {
+		deadlineTimer = time.NewTimer(currentDeadline.Sub(time.Now()))
```

Comment:
Are there any negative cases that should be handled here?

Reply:
Negative durations in a timer will tick immediately/send an event on the timer channel below. That seems safe to do here, it means that we are already past the deadline

Reply:
Yeah the negative case fires now which is desired if true